plugins/keep-markup/prism-keep-markup.js

```diff
@@ -38,14 +38,21 @@
 				return;
 			}
 
+			// Replace element with a clone
+			var clone = element.cloneNode(false);
+			element.parentNode.replaceChild(clone, element);
+			while (element.firstChild) {
+				clone.appendChild(element.firstChild);
+			}
```

Comment:
Why is this necessary?

From what I see, the `before-highlight` hook doesn't change the current DOM tree. (At least, it doesn't in a way that really matters to Prism.) We also know that inserting the kept DOM nodes happens [after insert](https://github.com/PrismJS/prism/blob/3bd8fdb15a24a8ce901a7690256752b168dd6288/components/prism-core.js#L590). Since insert will automatically detach the nodes we indent to keep, why is the cloning here necessary? Couldn't we leave the DOM tree unchanged in this hook, assume that the nodes have been detached in `after-highlight`, and then just use the original nodes without ever cloning them?

Reply:
@RunDevelopment I did this so that data contains only detached empty nodes. Later, when we're inserting the node back into the highlighted tree, its content needs to be discarded anyway, so we can [append the extracted range](https://github.com/PrismJS/prism/blob/d50d37c4bfcb92348fe67eba8c38c036aa31c85f/plugins/keep-markup/prism-keep-markup.js#L110).

We could store directly the element with its children, and only empty it just before we append the range. But we'd store unneeded information in data.

Reply:
> We could store directly the element with its children, and only empty it just before we append the range.

Yup, that's what I meant to suggest. This has the nice property that we don't need to do any cloning at all.

> But we'd store unneeded information in data.

Not that it's really a problem. Assuming that we keep all the markup in a code block, we would need to store all (element) nodes anyway. The only unneeded information is the tree structure between those nodes + the text, and that's just not that much in terms of memory.